pkg/app/piped/apistore/eventstore/store.go

```diff
@@ -0,0 +1,154 @@
+}
+
+type store struct {
+	apiClient    apiClient
+	syncInterval time.Duration
+	gracePeriod  time.Duration
+	logger       *zap.Logger
+
+	// Mark that it has handled all events that was created before this value.
+	milestone int64
+	// The key is supposed to be event-definition ID, a string consists of name and labels.
+	// And the value is the address to the latest Event.
+	latestEventMap atomic.Value
+}
+
+const (
+	defaultSyncInterval = time.Minute
+)
+
+// NewStore creates a new event store instance.
+// This syncs with the control plane to keep the list of events for this runner up-to-date.
+func NewStore(apiClient apiClient, gracePeriod time.Duration, logger *zap.Logger) Store {
+	s := &store{
+		apiClient:    apiClient,
+		syncInterval: defaultSyncInterval,
+		gracePeriod:  gracePeriod,
+		logger:       logger.Named("event-store"),
+	}
+	s.latestEventMap.Store(make(map[string]*model.Event))
+	return s
+}
+
+// Run starts runner that periodically makes the Events in the cache up-to-date
+// by fetching from the control-plane.
+func (s *store) Run(ctx context.Context) error {
+	s.logger.Info("start running event store")
+
+	syncTicker := time.NewTicker(s.syncInterval)
+	defer syncTicker.Stop()
+
+	// Do first sync without waiting the first ticker.
+	s.milestone = time.Now().Add(-time.Hour).Unix()
+	s.sync(ctx)
+
+	for {
+		select {
+		case <-syncTicker.C:
+			s.sync(ctx)
+
+		case <-ctx.Done():
+			s.logger.Info("event store has been stopped")
+			return nil
+		}
+	}
+}
+
+// sync fetches a list of events inside the range between own milestone and the current local time.
+// Only this function takes responsibility for updating the cache.
+func (s *store) sync(ctx context.Context) error {
+	// TODO: Use UTC and let control-plane convert it into the control-plane's Local time
+	// Unexpected behavior can be happened if Piped uses different timezone from the control-plane.
+	to := time.Now().Unix()
+	resp, err := s.apiClient.ListEvents(ctx, &pipedservice.ListEventsRequest{
+		From:  s.milestone,
+		To:    to,
+		Order: pipedservice.ListEventsOrder_FROM_OLDEST,
+	})
+	if err != nil {
+		s.logger.Error("failed to list events", zap.Error(err))
+		return err
+	}
+	// Set this time's "to" as the next time's "from".
+	s.milestone = to
+
+	// Make the cache up-to-date by traversing events sorted by oldest first.
+	eventMap := s.latestEventMap.Load().(map[string]*model.Event)
+	for _, event := range resp.Events {
+		id := model.EventDefinitionID(event.Name, event.Labels)
+		eventMap[id] = event
+	}
+	s.latestEventMap.Store(eventMap)
+
+	return nil
+}
+
+func (s *store) Getter() Getter {
+	return s
+}
+
+func (s *store) GetLatest(ctx context.Context, name string, labels map[string]string) (*model.Event, bool) {
+	eventMap := s.latestEventMap.Load().(map[string]*model.Event)
+	id := model.EventDefinitionID(name, labels)
+	event, ok := eventMap[id]
```

Comment:
One thing I want to ask about our event matching, in case we have an event 

```
name = foo
labels: { label-1: value-1, label-2: value-2}
```

That event will match the following case or not?

```
piVersion: pipecd.dev/v1beta1
kind: EventWatcher
spec:
  events:
    - name: foo
      labels:
           label-1: value-1
      replacements:
        - file: dev/app1/deployment.yaml
          yamlField: $.spec.template.spec.containers[0].image
```

Reply:
Or all labels must be specified to be matched.

Reply:
I totally assume exact matches are needed, but on second thought, seems like that case should be treated as a matching case. Let me think, just a moment please...

Reply:
@nghialv I'm still debating but which one do you prefer?

Reply:
If possible, the event should be matched when the name is equal AND the event's labels contain the specified ones.
(Because they are labels and users will have more options.)

So I think we can change the map to a new data structure like `map[event-name]ListOfEventsHaveSameName`.

Reply:
> But one thing we have to consider is what happens when multiple events match the specified label list.

So in order to avoid this complexity and avoid adding the dynamic label by users (e.g. version number),
I think we should not allow subset matching.

Name and all labels must be matched.

Reply:
My feeling is exactly the same as yours. Some people can be helped by subset matching, but too complicated to handle for us. Thank you for a great opportunity to re-think it and letting me know your cool thought!